File: 207-course-schedule.py
from typing import List
from collections import defaultdict

class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graphs = defaultdict(list)


        for courseNumber, dependency in prerequisites:
            graphs[courseNumber].append(dependency)

        states = [0] * numCourses
        
        def isValid(node: int):
            if states[node] == 1:
                return False
            if states[node] == 2:
                return True
            
            states[node] = 1

            for neighbor in graphs[node]:
                if not isValid(neighbor):
                    return False

            states[node] = 2
            return True
        
        for course in range(numCourses):
            if not isValid(course):
              return False
        
        return True
    

print(Solution().canFinish(2, [[1,0]]))

File: 207-course-schedule.ts
function canFinish(numCourses: number, prerequisites: number[][]): boolean {
  if (prerequisites.length === 0) {
    return true;
  }
  const graphs: Map<number, number[]> = new Map();

  for (let i = 0; i < numCourses; i++) {
    graphs.set(i, []);
  }

  for (let requisites of prerequisites) {
    graphs.get(requisites[1])?.push(requisites[0]);
  }

  let seen = new Array(numCourses).fill(0);
  const visited: number[] = [];

  function dfs(node: number) {
    if (seen[node] === 2) {
      return true;
    }
    if (seen[node] === 1) {
      return false;
    }
    seen[node] = 1;

    for (const curr of graphs.get(node)!) {
      if (!dfs(curr)) {
        return false;
      }
    }

    seen[node] = 2;
    visited.push(node);

    return true;
  }

  for (let i = 0; i < numCourses; i++) {
    if (!dfs(i)) {
      return false;
    }
  }

  return true;
}

console.log(canFinish(2, [[0, 1]]));
